File: public/sample_codes/sample.py
from collections import defaultdict

#!-------------------- FORMAT HELPERS ----------------------------
br = "\\n"   # for newline

def getVectorToString(v):
    return "[" + ", ".join(map(str, v)) + "]"

def get2DVectorToString(v):
    return "[" + "\\n".join(getVectorToString(row) for row in v) + "]"  

#!---------------- YOUR RECURSIVE FUNCTION ------------------------

def fib(x, ok=0):
    nodeVal = f"fib(x = {x})"
    ok = tv.make(nodeVal, ok)

    if x < 2:
        tv.onHover(str(x), ok)
        return x

    tmp = fib(x - 1, ok) + fib(x - 2, ok)
    tv.onHover(str(tmp), ok)
    return tmp

#!---------------------- MAIN FUNCTION ----------------------------

def run_main():
    # Call your recursive function here
    n = 5
    fib(n)

    # Don't remove the below lines
    tv.buildTree()
    tv.printData()

#!-------------------- HELPER CLASS ------------------------- 

class TreeVisualizer:
    def __init__(self):
        self.nodeCount = 0
        self.nodeToString = {}
        self.nodeToId = {}
        self.nodeToAns = {}
        self.pos = defaultdict(lambda: [0.0, 0.0])
        self.modifier = defaultdict(float)
        self.adj = defaultdict(list)

    def make(self, nodeVal, parent_node):
        self.nodeCount += 1
        new_node = self.nodeCount
        self.nodeToString[new_node] = nodeVal
        self.nodeToId[nodeVal] = new_node
        self.adj[parent_node].append(new_node)
        return new_node

    def onHover(self, nodeVal, node):
        self.nodeToAns[node] = nodeVal

    def buildTree(self):
        for i in range(1, self.nodeCount + 1):
            if not self.adj[i]:
                continue
            for j in range(len(self.adj[i])):
                self.pos[self.adj[i][j]] = [float(j), 0.0]

        for i in range(1, self.nodeCount + 1):
            if not self.adj[i]:
                continue
            first_child = self.adj[i][0]
            last_child = self.adj[i][-1]
            mid = self.pos[i][0] - (self.pos[first_child][0] + self.pos[last_child][0]) / 2.0
            self.modifier[i] = mid

        def updateXY(root, depth, modSum):
            self.pos[root][0] += modSum
            self.pos[root][1] = depth
            for child in self.adj[root]:
                updateXY(child, depth + 1, modSum + self.modifier[root])

        updateXY(1, 0, 0.0)

        minPos = 0.0
        for coords in self.pos.values():
            minPos = min(minPos, coords[0])
        if minPos < 0:
            for k in self.pos:
                self.pos[k][0] -= minPos

        def getRB(node):
            curr = node
            while self.adj[curr]:
                curr = self.adj[curr][-1]
            return self.pos[curr][0]

        def getLB(node):
            curr = node
            while self.adj[curr]:
                curr = self.adj[curr][0]
            return self.pos[curr][0]

        def ShiftSubtree(node, shift):
            self.pos[node][0] += shift
            for child in self.adj[node]:
                ShiftSubtree(child, shift)

        def fixNodeConflicts(root):
            for child in self.adj[root]:
                fixNodeConflicts(child)
            if not self.adj[root]:
                return
            for j in range(len(self.adj[root]) - 1):
                child1 = self.adj[root][j]
                child2 = self.adj[root][j + 1]
                RB_ofLeft = getRB(child1)
                LB_ofRight = getLB(child2)
                if RB_ofLeft >= LB_ofRight:
                    ShiftSubtree(child2, RB_ofLeft - LB_ofRight + 1.0)
            
            first_child = self.adj[root][0]
            last_child = self.adj[root][-1]
            self.pos[root][0] = (self.pos[first_child][0] + self.pos[last_child][0]) / 2.0

        fixNodeConflicts(1)

    def printData(self):
        print("{")
        print('"nodes": [')
        nodes_items = list(self.nodeToString.items())
        for idx, (node_id, label) in enumerate(nodes_items):
            x = self.pos[node_id][0]
            y = self.pos[node_id][1]
            hidden = self.nodeToAns.get(node_id, "")
            line = f'{{ "id": "{node_id}", "position": {{"x": {x}, "y": {y}}}, "data": {{ "label": "{label}", "hidden": "{hidden}"}} }}'
            if idx != len(nodes_items) - 1:
                line += ","
            print(line)
        print("],")
        print('"edges": [')
        tot = 1
        edges = []
        for i in range(1, self.nodeCount + 1):
            for x in self.adj[i]:
                edges.append((i, x))
        
        for idx, (src, tgt) in enumerate(edges):
            line = f'{{ "id": "{tot}", "source": "{src}", "target" : "{tgt}" }}'
            if idx != len(edges) - 1:
                line += ","
            print(line)
            tot += 1
        print("]")
        print("}")

#!----------------------- INITIALIZATION & EXECUTION TRIGGER -----------------------

tv = TreeVisualizer()

if __name__ == "__main__":
    run_main()

File: public/sample_codes/sample.js
// !-------------------- FORMAT HELPERS ----------------------------

const br = "\\n";  // Use this for new lines

function getVectorToString(v) {
    return "[" + v.join(", ") + "]";
} 

function get2DVectorToString(v) {
    return "[" + v.map(row => getVectorToString(row)).join("\\n") + "]";
}

// !-------------------- YOUR RECURSIVE FUNCTION & MAIN ----------------------------

function fib(x, ok = 0) {
    const nodeVal = `fib(x = ${x})`;
    ok = tv.make(nodeVal, ok);

    if (x < 2) {
        tv.onHover(String(x), ok);
        return x;
    }

    const tmp = fib(x - 1, ok) + fib(x - 2, ok);
    tv.onHover(String(tmp), ok);
    return tmp;
}

function runMain() {
    // Call your recursive function here
    const n = 5;
    fib(n);

    // Don't remove the below lines
    tv.buildTree();
    tv.printData();
}

// !-------------------- HELPER CLASSES (HIDDEN AT BOTTOM) -------------------------

class TreeVisualizer {
    constructor() {
        this.nodeCount = 0;
        this.nodeToString = {};
        this.nodeToId = {};
        this.nodeToAns = {};
        this.pos = {};
        this.modifier = {};
        this.adj = {};
    }

    // Helpers to mimic Python's defaultdict behavior
    getAdj(node) {
        if (!this.adj[node]) this.adj[node] = [];
        return this.adj[node];
    }

    getPos(node) {
        if (!this.pos[node]) this.pos[node] = [0.0, 0.0];
        return this.pos[node];
    }

    getModifier(node) {
        if (this.modifier[node] === undefined) this.modifier[node] = 0.0;
        return this.modifier[node];
    }

    make(nodeVal, parentNode) {
        this.nodeCount += 1;
        const newNode = this.nodeCount;
        this.nodeToString[newNode] = nodeVal;
        this.nodeToId[nodeVal] = newNode;
        this.getAdj(parentNode).push(newNode);
        return newNode;
    }

    onHover(nodeVal, node) {
        this.nodeToAns[node] = nodeVal;
    }

    buildTree() {
        for (let i = 1; i <= this.nodeCount; i++) {
            const children = this.getAdj(i);
            if (children.length === 0) continue;
            for (let j = 0; j < children.length; j++) {
                this.getPos(children[j])[0] = parseFloat(j);
            }
        }

        for (let i = 1; i <= this.nodeCount; i++) {
            const children = this.getAdj(i);
            if (children.length === 0) continue;
            const firstChild = children[0];
            const lastChild = children[children.length - 1];
            const mid = this.getPos(i)[0] - (this.getPos(firstChild)[0] + this.getPos(lastChild)[0]) / 2.0;
            this.modifier[i] = mid;
        }

        const updateXY = (root, depth, modSum) => {
            this.getPos(root)[0] += modSum;
            this.getPos(root)[1] = depth;
            for (const child of this.getAdj(root)) {
                updateXY(child, depth + 1, modSum + this.getModifier(root));
            }
        };

        updateXY(1, 0, 0.0);

        let minPos = 0.0;
        for (const coords of Object.values(this.pos)) {
            minPos = Math.min(minPos, coords[0]);
        }
        if (minPos < 0) {
            for (const k in this.pos) {
                this.pos[k][0] -= minPos;
            }
        }

        const getRB = (node) => {
            let curr = node;
            while (this.getAdj(curr).length > 0) {
                const children = this.getAdj(curr);
                curr = children[children.length - 1];
            }
            return this.getPos(curr)[0];
        };

        const getLB = (node) => {
            let curr = node;
            while (this.getAdj(curr).length > 0) {
                curr = this.getAdj(curr)[0];
            }
            return this.getPos(curr)[0];
        };

        const shiftSubtree = (node, shift) => {
            this.getPos(node)[0] += shift;
            for (const child of this.getAdj(node)) {
                shiftSubtree(child, shift);
            }
        };

        const fixNodeConflicts = (root) => {
            const children = this.getAdj(root);
            for (const child of children) {
                fixNodeConflicts(child);
            }
            if (children.length === 0) return;

            for (let j = 0; j < children.length - 1; j++) {
                const child1 = children[j];
                const child2 = children[j + 1];
                const rbOfLeft = getRB(child1);
                const lbOfRight = getLB(child2);
                if (rbOfLeft >= lbOfRight) {
                    shiftSubtree(child2, rbOfLeft - lbOfRight + 1.0);
                }
            }

            const firstChild = children[0];
            const lastChild = children[children.length - 1];
            this.getPos(root)[0] = (this.getPos(firstChild)[0] + this.getPos(lastChild)[0]) / 2.0;
        };

        fixNodeConflicts(1);
    }

    printData() {
        console.log("{");
        console.log('  "nodes": [');
        const nodesItems = Object.entries(this.nodeToString);
        
        nodesItems.forEach(([nodeId, label], idx) => {
            const x = this.getPos(nodeId)[0];
            const y = this.getPos(nodeId)[1];
            const hidden = this.nodeToAns[nodeId] || "";
            let line = `    { "id": "${nodeId}", "position": {"x": ${x}, "y": ${y}}, "data": { "label": "${label}", "hidden": "${hidden}"} }`;
            if (idx !== nodesItems.length - 1) {
                line += ",";
            }
            console.log(line);
        });
        
        console.log("  ],");
        console.log('  "edges": [');
        
        let tot = 1;
        const edges = [];
        for (let i = 1; i <= this.nodeCount; i++) {
            for (const x of this.getAdj(i)) {
                edges.push([i, x]);
            }
        }

        edges.forEach(([src, tgt], idx) => {
            let line = `    { "id": "${tot}", "source": "${src}", "target" : "${tgt}" }`;
            if (idx !== edges.length - 1) {
                line += ",";
            }
            console.log(line);
            tot += 1;
        });
        
        console.log("  ]");
        console.log("}");
    }
}

//! ---------------------INITIALIZATION & EXECUTION TRIGGER-------------------------

const tv = new TreeVisualizer();

// Standard JS environment execution block equivalent to `if __name__ == "__main__":`
if (typeof require !== 'undefined' && require.main === module) {
    runMain();
} else if (typeof window !== 'undefined') {
    // Fallback if running directly in a browser
    runMain();
} else {
    // Fallback for general execution like quick JS playgrounds
    runMain();
}